File: nodejs/src/core/registry.ts
import * as path from 'path';
import { BaseParser, ParserData } from '../parsers/base';
import { JSONParser } from '../parsers/json_parser';
import { CSVParser } from '../parsers/csv_parser';
import { NginxParser } from '../parsers/nginx_parser';

export interface FormatInfo {
  name: string;
  displayName: string;
  extensions: string[];
  mimeTypes: string[];
}

export class ParserRegistry {
  private parsers: Map<string, typeof BaseParser> = new Map();
  private extensionMap: Map<string, string> = new Map();

  constructor() {
    this.register(JSONParser);
    this.register(CSVParser);
    this.register(NginxParser);
  }

  register(parserClass: any): void {
    const parser = new parserClass() as BaseParser;
    const formatName = parser.formatName;

    this.parsers.set(formatName, parserClass);

    for (const ext of parser.fileExtensions) {
      const normalizedExt = ext.startsWith('.') ? ext.toLowerCase() : '.' + ext.toLowerCase();
      this.extensionMap.set(normalizedExt, formatName);
    }
  }

  getParser(formatName: string): BaseParser | null {
    const parserClass = this.parsers.get(formatName);
    return parserClass ? new parserClass() as BaseParser : null;
  }

  detectFormat(filePath: string): string | null {
    const ext = path.extname(filePath).toLowerCase();
    return this.extensionMap.get(ext) || null;
  }

  findParserForFile(filePath: string, formatHint?: string): BaseParser | null {
    if (formatHint) {
      const parser = this.getParser(formatHint);
      if (parser && parser.supportsFormat(filePath)) {
        return parser;
      }
    }

    const formatName = this.detectFormat(filePath);
    if (formatName) {
      const parser = this.getParser(formatName);
      if (parser && parser.supportsFormat(filePath)) {
        return parser;
      }
    }

    for (const parserClass of this.parsers.values()) {
      const parser = new parserClass() as BaseParser;
      if (parser.supportsFormat(filePath)) {
        return parser;
      }
    }

    return null;
  }

  listSupportedFormats(): FormatInfo[] {
    const formats: FormatInfo[] = [];
    for (const [formatName, parserClass] of this.parsers) {
      const parser = new parserClass() as BaseParser;
      formats.push({
        name: formatName,
        displayName: parser.getDisplayName(),
        extensions: parser.fileExtensions,
        mimeTypes: parser.mimeTypes
      });
    }
    return formats;
  }
}

export const registry = new ParserRegistry();

File: core/registry.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Plugin registry for file format parsers
"""

from typing import Dict, List, Optional, Type
from pathlib import Path


class ParserRegistry:
    """Central registry for format parsers"""

    def __init__(self):
        self._parsers: Dict[str, Type['BaseParser']] = {}
        self._extension_map: Dict[str, str] = {}

    def register(self, parser_class: Type['BaseParser']) -> None:
        """
        Register a parser class.

        Args:
            parser_class: Parser class to register
        """
        parser = parser_class()
        format_name = parser.format_name

        self._parsers[format_name] = parser_class

        # Build extension map for auto-detection
        for ext in parser.file_extensions:
            # Normalize extension (with or without leading dot)
            if not ext.startswith('.'):
                ext = '.' + ext
            self._extension_map[ext.lower()] = format_name

    def get_parser(self, format_name: str) -> Optional['BaseParser']:
        """Get parser instance by format name"""
        parser_class = self._parsers.get(format_name)
        return parser_class() if parser_class else None

    def detect_format(self, file_path: str) -> Optional[str]:
        """
        Auto-detect format based on file extension.

        Args:
            file_path: Path to the file

        Returns:
            Format name or None if not detected
        """
        ext = Path(file_path).suffix.lower()
        return self._extension_map.get(ext)

    def find_parser_for_file(self, file_path: str, format_hint: Optional[str] = None) -> Optional['BaseParser']:
        """
        Find appropriate parser for a file.

        Args:
            file_path: Path to the file
            format_hint: Optional format override

        Returns:
            Parser instance or None if no suitable parser found
        """
        # Use manual override if provided
        if format_hint:
            parser = self.get_parser(format_hint)
            if parser and parser.supports_format(file_path):
                return parser

        # Try auto-detection
        format_name = self.detect_format(file_path)
        if format_name:
            parser = self.get_parser(format_name)
            if parser and parser.supports_format(file_path):
                return parser

        # Fallback: try all registered parsers
        for parser_class in self._parsers.values():
            parser = parser_class()
            if parser.supports_format(file_path):
                return parser

        return None

    def list_supported_formats(self) -> List[Dict[str, any]]:
        """Return list of all registered formats with metadata"""
        formats = []
        for format_name, parser_class in self._parsers.items():
            parser = parser_class()
            formats.append({
                'name': format_name,
                'display_name': parser.get_display_name(),
                'extensions': parser.file_extensions,
                'mime_types': parser.mime_types
            })
        return formats


# Global registry instance
registry = ParserRegistry()
